cpp/src/arrow/util/thread_pool.h

```diff
@@ -605,15 +620,7 @@ typename Fut::SyncType RunSynchronously(FnOnce<Fut(Executor*)> get_future,
 template <typename T>
 Iterator<T> IterateSynchronously(
     FnOnce<Result<std::function<Future<T>()>>(Executor*)> get_gen, bool use_threads) {
-  if (use_threads) {
-    auto maybe_gen = std::move(get_gen)(GetCpuThreadPool());
-    if (!maybe_gen.ok()) {
-      return MakeErrorIterator<T>(maybe_gen.status());
-    }
-    return MakeGeneratorIterator(*maybe_gen);
-  } else {
-    return SerialExecutor::IterateGenerator(std::move(get_gen));
-  }
+  return IterateSynchronously(std::move(get_gen), use_threads, GetCpuThreadPool());
```

Comment:
This implicitly introducing a behavior change, previously if `!use_threads`, `GetCpuThreadPool()` would be nevercall, and maybe the global cpu pool would not being initialized. However, seems here it would be initialized?

Reply:
Good point. I modified the API to accept nulls and if it is null, we use the default CPU thread pool, but only in `use_threads == true` case